File: vmware_perl/vmware_set_path_policy.pl
#!/usr/bin/perl
use strict;
use VMware::VIRuntime;
use libsf;
use libvmware;
use JSON::XS;

use Data::Dumper;

# Set default username/password to use
# These can be overridden via --username and --password command line options
Opts::set_option("username", "administrator");
Opts::set_option("password", "password");

# Set default vCenter Server
# This can be overridden with --mgmt_server
Opts::set_option("server", "vcenter.domain.local");

my %opts = (
    mgmt_server => {
        type => "=s",
        help => "The hostname/IP of the vCenter Server (replaces --server)",
        required => 0,
        default => Opts::get_option("server"),
    },
    vmhost => {
        type => "=s",
        help => "The hostname/IP of the host to create datastores on",
        required => 1,
    },
    policy => {
        type => "=s",
        help => "The policy to set (fixed/rr/mru)",
        required => 1,
    },
    result_address => {
        type => "=s",
        help => "Address of a ZMQ server listening for results (when run as a child process)",
        required => 0,
    },
    debug => {
        type => "",
        help => "Display more verbose messages",
        required => 0,
    },
);

Opts::add_options(%opts);
if (scalar(@ARGV) < 1)
{
   print "Set the path selection policy for the LUNS on a particular host\n";
   Opts::usage();
   exit 1;
}
Opts::parse();
my $vsphere_server = Opts::get_option("mgmt_server");
Opts::set_option("server", $vsphere_server);
my $host_name = Opts::get_option('vmhost');
my $policy = Opts::get_option('policy');
my $enable_debug = Opts::get_option('debug');
my $result_address = Opts::get_option('result_address');
Opts::validate();

if ($policy !~ /fixed/i &&
    $policy !~ /rr/i &&
    $policy !~ /mru/)
{
    print "Please provide one of fixed/rr/mru for policy\n";
    exit 1;
}
$policy = "VMW_PSP_FIXED" if ($policy =~ /fixed/i);
$policy = "VMW_PSP_RR" if ($policy =~ /rr/i);
$policy = "VMW_PSP_MRU" if ($policy =~ /mru/i);

# Turn on debug events if requested
$mylog::DisplayDebug = 1 if $enable_debug;

# Turn off cert validation so we can get away with self signed certs
mylog::debug("Disabling SSL cert verification");
$ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;

# Connect to vSphere
mylog::info("Connecting to vSphere at $vsphere_server...");
eval
{
   Util::connect();
};
if ($@)
{
   mylog::error("Could not connect to $vsphere_server: $!");
   exit 1;
}

# Find the host
mylog::info("Searching for host $host_name");
my $vmhost = Vim::find_entity_view(view_type => 'HostSystem', filter => {'name' => qr/^$host_name$/i}, properties => ['configManager', 'config']);
if (!$vmhost)
{
    mylog::error("Could not find host '$host_name'");
    exit 1;
}
my $storage_manager = Vim::get_view(mo_ref => $vmhost->configManager->storageSystem);
#my $a = $storage_manager->QueryPathSelectionPolicyOptions();
#print Dumper($a) . "\n";
#exit 1;

# Convert LUN to cannonical name
my %lun2name;
foreach my $lun (@{$vmhost->config->storageDevice->scsiLun})
{
    $lun2name{$lun->key} = $lun->canonicalName;
}

# Find the SolidFire disks and set the multipath policy
my $allgood = 1;
foreach my $lun (@{$storage_manager->storageDeviceInfo->multipathInfo->lun})
{
    my $name = $lun2name{$lun->lun};
    # Skip non-SF devices
    if ($name !~ /f47acc/)
    {
        mylog::debug("Skipping $name");
        next;
    }
    
#    print Dumper($lun->policy) . "\n";
    if ($lun->policy->policy =~ $policy)
    {
        mylog::debug("$name is already using policy $policy");
        next;
    }

    mylog::debug("Setting path policy on $name");
#    print Dumper(HostMultipathInfoLogicalUnitPolicy->new(policy => $policy)) . "\n";
    eval
    {
        $storage_manager->SetMultipathLunPolicy(lunId => $lun->id,
                                                policy => HostMultipathInfoLogicalUnitPolicy->new(policy => $policy));
    };
    if ($@)
    {
        my $fault = $@;
        print Dumper($fault);
        libvmware::DisplayFault("Setting path policy on $name failed", $fault);
        $allgood = 0;
    }
    last;
}

if ($allgood)
{
    mylog::pass("Sucessfully set path policy on $host_name");
}
else
{
    mylog::error("Failed to set path policy on all volumes");
}

# Send the info back to parent script if requested
if (defined $result_address)
{
    libsf::SendResultToParent(result_address => $result_address, result => $allgood);
}
exit $allgood;

File: vmware_set_path_policy.py
#!/usr/bin/env python

"""
This action will set the multipathing policy (load balancing policy) on the LUNs attached to a VMware host

When run as a script, the following options/env variables apply:
    --mgmt_server       The IP/hostname of the vSphere Server

    --mgmt_user         The vsphere admin username

    --mgmt_pass         The vsphere admin password

    --vmhost            The IP of the ESX host to set the policy on

    --policy            The new path selection policy to set
"""

import sys
from optparse import OptionParser
import lib.libsf as libsf
from lib.libsf import mylog
import lib.sfdefaults as sfdefaults
import lib.libsfcluster as libsfcluster
from lib.action_base import ActionBase
import lib.libvmware as libvmware

class VmwareSetPathPolicyAction(ActionBase):
    class Events:
        """
        Events that this action defines
        """

    def __init__(self):
        super(self.__class__, self).__init__(self.__class__.Events)

    def ValidateArgs(self, args):
        libsf.ValidateArgs({"mgmt_server" : libsf.IsValidIpv4Address,
                            "mgmt_user" : None,
                            "mgmt_pass" : None,
                            "vmhost" : None,
                            "new_policy" : None},
            args)

    def Execute(self, vmhost, new_policy, mgmt_server=sfdefaults.fc_mgmt_server, mgmt_user=sfdefaults.fc_vsphere_user, mgmt_pass=sfdefaults.fc_vsphere_pass, debug=False):
        """
        Set the multipath policy
        """
        self.ValidateArgs(locals())
        if debug:
            mylog.showDebug()
        else:
            mylog.hideDebug()

        if new_policy.lower() == "rr":
            new_policy = "VMW_PSP_RR"
        elif new_policy.lower() == "fixed":
            new_policy = "VMW_PSP_FIXED"
        elif new_policy.lower() == "mru":
            new_policy = "VMW_PSP_MRU"

        mylog.info("Connecting to vSphere " + mgmt_server)
        try:
            with libvmware.VsphereConnection(mgmt_server, mgmt_user, mgmt_pass) as vsphere:
                # Find the requested host
                mylog.info("Searching for host " + vmhost)
                host = libvmware.FindHost(vsphere, vmhost)

                storage_sys = host.configManager.storageSystem

                mylog.info("Checking path selection policies")
                policies = storage_sys.QueryPathSelectionPolicyOptions()
                allowed_policies = [p.policy.key for p in policies]
                if new_policy not in allowed_policies:
                    mylog.error("Unrecognized multipath policy: only " + " / ".join(allowed_policies) + " are allowed")
                    return False

                mylog.info("Getting a list of LUNs")
                lun2name = {}
                for lun in host.config.storageDevice.scsiLun:
                    lun2name[lun.key] = lun.canonicalName

                allgood = True
                for lun in storage_sys.storageDeviceInfo.multipathInfo.lun:
                    name = lun2name[lun.lun]
                    # Skip non-SF devices
                    if "f47acc" not in name:
                        continue

                    if lun.policy.policy == new_policy:
                        mylog.info(name + " is already using policy " + new_policy)
                        continue

                    mylog.info("Setting " + name + " to " + new_policy)
                    if new_policy != "VMW_PSP_FIXED":
                        p = vim.host.MultipathInfo.LogicalUnitPolicy()
                        p.policy = new_policy
                        try:
                            storage_sys.SetMultipathLunPolicy(lun.id, p)
                        except vmodl.MethodFault as e:
                            mylog.error("Failed to set multipath policy on " + name + ": " + str(e))
                            allgood = False
                    else:
                        # Hacky, but it appears the only workable way to set the policy to fixed is through esxcli or powercli.  WTF
                        ssh = libsf.ConnectSsh(vmhost, "root", "solidfire")
                        stdin, stdout, stderr = libsf.ExecSshCommand(ssh, "esxcli storage nmp device set --device=\"" + name + "\" --psp=VMW_PSP_FIXED")
                        output = "\n".join(stdout.readlines()).strip()
                        err = "\n".join(stderr.readlines()).strip()
                        if output or err:
                            mylog.error("Failed to set multipath policy on " + name + ": " + output + err)
                            allgood = False
                        else:
                            # Wait for the policy to change
                            stdin, stdout, stderr = libsf.ExecSshCommand(ssh, "POLICY=; while [[ \"$POLICY\" != \"" + new_policy + "\" ]]; do sleep 1; POLICY=$(esxcli storage nmp device list | grep -A8 " + name + " | grep 'Path Selection Policy:' | awk '{print $4}'); if [[ $? != \"0\" ]]; then break; fi; done")
                        ssh.close()
        except libvmware.VmwareError as e:
            mylog.error(str(e))
            return False

        if allgood:
            mylog.passed("Successfully set multipath policy on " + vmhost)
            return True
        else:
            mylog.error("Failed to set multipath policy on all volumes")
            return False


# Instantate the class and add its attributes to the module
# This allows it to be executed simply as module_name.Execute
libsf.PopulateActionModule(sys.modules[__name__])

if __name__ == '__main__':
    mylog.debug("Starting " + str(sys.argv))

    # Parse command line arguments
    parser = OptionParser(option_class=libsf.ListOption, description=libsf.GetFirstLine(sys.modules[__name__].__doc__))
    parser.add_option("-s", "--mgmt_server", type="string", dest="mgmt_server", default=sfdefaults.fc_mgmt_server, help="the IP/hostname of the vSphere Server [%default]")
    parser.add_option("-m", "--mgmt_user", type="string", dest="mgmt_user", default=sfdefaults.fc_vsphere_user, help="the vsphere admin username [%default]")
    parser.add_option("-a", "--mgmt_pass", type="string", dest="mgmt_pass", default=sfdefaults.fc_vsphere_pass, help="the vsphere admin password [%default]")
    parser.add_option("-o", "--vmhost", type="string", dest="vmhost", default=None, help="the IP of the ESX host to set the policy on")
    parser.add_option("--policy", type="string", dest="policy", default=None, help="the new path selection policy to set")
    parser.add_option("--debug", action="store_true", dest="debug", default=False, help="display more verbose messages")
    (options, extra_args) = parser.parse_args()

    try:
        timer = libsf.ScriptTimer()
        if Execute(vmhost=options.vmhost, new_policy=options.policy, mgmt_server=options.mgmt_server, mgmt_user=options.mgmt_user, mgmt_pass=options.mgmt_pass, debug=options.debug):
            sys.exit(0)
        else:
            sys.exit(1)
    except libsf.SfArgumentError as e:
        mylog.error("Invalid arguments - \n" + str(e))
        sys.exit(1)
    except SystemExit:
        raise
    except KeyboardInterrupt:
        mylog.warning("Aborted by user")
        Abort()
        sys.exit(1)
    except:
        mylog.exception("Unhandled exception")
        sys.exit(1)
